### DISCUSSION TITLE
Re-compiling with new version

### DISCUSSION THREAD

**Discussion by Woodcat64** [Jyers/Marlin#151] [Q&A] [answered]
I compiled v1.1.1. using config templates plus some changes I made. 

Now, when new version comes, what is the process to upgrade? Do I have to download the new version zip and overwrite everything, use new templates + my changes again? Or I keep the config files and overwrite just the src and lib directories in Marlin dir?

**Comment by clintiepoo** [comment, marked as answer]
This is where git can make your life easier. Here's what I did:
- Pulled the marlin branch from here
- created my own branch locally in VSCode
- made my "local" changes, commit as necessary to snapshot my local changes
- As new changes come out, I do a git pull
- Merge the changes into my local branch. If there are conflicts, I evaluate which I want keep.

**Reply by darco415** [reply]
nice!...so new to this....so right now I see 4 pull request here...so if I choose to pull say UBL, that will integrate into my local branch? ...ok I think im getting confused with commits...I was hoping to pull a commit into my local branch w/o having to d/l the entire zip folder...

**Reply by clintiepoo** [reply]
Those pull requests are people who have made changes and they are basically asking for them to be integrated into Jyers' repo. That's not what we're talking about.

To do what I'm saying, you'll need to download the repo again I think, but this time you'll want to clone it. 
- In VS Code, you can "clone repository" - clone this [repo](https://github.com/Jyers/Marlin.git). Put it in a new folder, away from your other changes. 
- Create new branch. I called mine "my_branch"
- Integrate your local changes. Commit to my_branch.
- As Jyers makes changes, switch back to the Ender_3_V2_Extensible_UI branch and do a "pull" to get his changes.
- Do a branch merge to integrate his changes in with yours.

This is honestly a little hard at first, at least it was for me. [Git Extensions](http://gitextensions.github.io/) makes it much easier to visualize what's going on and see exactly what files are changing. Good luck!

**Reply by darco415** [reply]
ok...I think im good now...I did d/l the github Desktop which makes things easier..so I now need to wait for Mr. Meyers to make some new commits...thxs bro

**Reply by Woodcat64** [reply]
Thanks, I will give it a try.

**Reply by Schnabulation** [reply]
@clintiepoo Can I suggest you something? Instead of creating a local branch, you could create a „fork“ of Jyers repository. This acts as YOUR OWN repository on which you have full control on. It is even uploaded to GitHub with your personal changes. So when your local folder gets deleted somehow you just pull your personal fork again.
Also if Jyers makes changes you just merge his changes into your own fork - that‘s how I do it.

**Reply by Schnabulation** [reply]
Look, this is my fork of his firmware - I have full control over this repo and can even push my commits to GitHub. It‘s basically his firmware but with my config-files: https://github.com/Schnabulation/Marlin

**Reply by darco415** [reply]
When Im ready to push commits, which I doubt I ever will be, Ill just stick to the local branch.

**Reply by Woodcat64** [reply]
This is not a rant, but I give up. 
Given that all I wanted is to build and install cloud free smart garage opener based on ESPhome and using Home Assistant and needed a custom enclosure for it, I think I went too far into this rabbit hole. :-) Now VS code won't even let me to compile, LOL. 

From the pull requests I see that some amazing things are coming. So, I'm just gonna download the source when ever new version comes, add my changes.

Thanks for help

**Reply by clintiepoo** [reply]
> 
> 
> @clintiepoo Can I suggest you something? Instead of creating a local branch, you could create a „fork“ of Jyers repository. This acts as YOUR OWN repository on which you have full control on. It is even uploaded to GitHub with your personal changes. So when your local folder gets deleted somehow you just pull your personal fork again.
> Also if Jyers makes changes you just merge his changes into your own fork - that‘s how I do it.

Thank you - I may do this. I'm still learning git. I actually am self-hosting gitlab at home and just pushing there, so I can backup my own changes. I'll consider forking, but for now I'm happy like this I think.

**Reply by swbog** [reply]
Do you guys use git bash to merge your forks after changes? Or is there a way to use git desktop. Bash is a PITA

**Reply by Schnabulation** [reply]
I'm using Git Bash:

`git fetch upstream`
`git merge upstream/Ender_3_V2_Extensible_UI`
`git push`

...that's it!

**Reply by swbog** [reply]
Thank you. That’s what I did was kind of annoying to set up the bash. I’ve never had an install ask for that many options I just clicked next through them all. I also expected it to ask me which changes to submit. But I suppose it’ll do that if there are conflicts maybe
